File: Source/InputValidator.m
#import "InputValidator.h"
#import "NumberInputValidator.h"

@interface InputValidator ()

@property (nonatomic) Validation *validation;

@end

@implementation InputValidator

- (instancetype)initWithValidation:(Validation *)validation {
    self = [super init];
    if (self) {
        _validation = validation;
    }

    return self;
}

- (BOOL)validateString:(NSString *)string {
    return [self validateReplacementString:nil withText:string withRange:NSMakeRange(0, 0)];
}

- (BOOL)validateReplacementString:(NSString *)replacementString withText:(NSString *)text withRange:(NSRange)range {
    BOOL shouldSkipValidations = (!self.validation);
    if (shouldSkipValidations) {
        return YES;
    }

    NSString *evaluatedString = text;

    BOOL valid = YES;

    if (self.validation.required == YES && !self.validation.minimumLength) {
        self.validation.minimumLength = @1;
    }

    if (self.validation.maximumLength || self.validation.minimumLength) {
        NSUInteger textLength = text.length;

        if (replacementString.length > 0) {
            textLength += replacementString.length;
        }

        if (self.validation.maximumLength) {
            valid = (textLength <= [self.validation.maximumLength unsignedIntegerValue]);
        }

        if (valid && self.validation.minimumLength) {
            valid = (textLength >= [self.validation.minimumLength unsignedIntegerValue]);
        }
    }

    if (replacementString) {
        NSMutableString *composedString = [[NSMutableString alloc] initWithString:text];
        [composedString insertString:replacementString atIndex:range.location];
        evaluatedString = [composedString copy];
    }

    if (text && valid && (self.validation.maximumValue || self.validation.minimumValue)) {
        NSNumberFormatter *formatter = [NSNumberFormatter new];
        formatter.locale = [NSLocale localeWithLocaleIdentifier:@"en_US"];
        NSNumber *newValue = [formatter numberFromString:evaluatedString];

        if (newValue) {
            if (self.validation.maximumValue) {
                valid = ([newValue floatValue] <= [self.validation.maximumValue floatValue]);
            }

            if (valid && self.validation.minimumValue) {
                valid = ([newValue floatValue] >= [self.validation.minimumValue floatValue]);
            }
        }
    }

    if (valid && self.validation.format) {
        NSError *formattingError = nil;
        valid = [self validateString:evaluatedString withFormat:self.validation.format error:&formattingError];
        if (formattingError) {
            NSLog(@"There was a problem checking the format of %@ (Format: %@, Error: %@)", evaluatedString, self.validation.format, formattingError);
        }
    }

    return valid;
}

- (BOOL)validateString:(NSString *)fieldValue withFormat:(NSString *)format error:(NSError **)error {
    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:format options:NSRegularExpressionCaseInsensitive error:error];
    NSRange range = [regex rangeOfFirstMatchInString:fieldValue options:NSMatchingReportProgress range:NSMakeRange(0, fieldValue.length)];

    return (range.location == 0 && range.length == fieldValue.length);
}

@end

File: Source/InputValidator.swift
import Foundation
import Validation

public class InputValidator {
    public var validation: Validation

    public init(validation: Validation) {
        self.validation = validation
    }

    public func validateString(string: String) -> Bool {
        return self.validation.validateString(string)
    }

    // This method is useful for partial validations, or validations where the final string is
    // in process of been completed. For example when entering characters into an UITextField.
    public func validateReplacementString(replacementString: String?, fullString: String?, inRange range: NSRange?) -> Bool {
        let evaluatedString = self.composedString(replacementString, fullString: fullString, inRange: range)

        return self.validation.validateString(evaluatedString)
    }

    public func composedString(replacementString: String?, fullString: String?, inRange range: NSRange?) -> String {
        var composedString = fullString ?? ""

        if let replacementString = replacementString, range = range {
            let index = composedString.startIndex.advancedBy(range.location)
            if range.location == composedString.characters.count {
                composedString.insertContentsOf(replacementString.characters, at: index)
            } else {
                composedString = (composedString as NSString).stringByReplacingCharactersInRange(range, withString: replacementString)
            }
            return composedString
        }

        return composedString
    }
}
